### ISSUE TITLE
Why discarding other token's logits (like `obs_text's logits`) when calculating the `action_logits`?

### ISSUE THREAD

**Issue by Felixvillas** [WeihaoTan/TWOSOME#4]
Hi~ Another Question
I found the following code in the `get_action_and_value` of `overcooked policy_pomdp`:
```python
action_list = [item for sublist in action_list for item in sublist]
self.action_list_ids = self.tokenizer(action_list, return_tensors="pt", padding=True)

self.action_list_length = torch.sum(self.action_list_ids["attention_mask"], dim = -1) - 1 #delete first token
sequence_length = torch.sum(attention_mask, dim = -1)
# find the [acion_start_index, end_start_index] in the sequence. sequence is like "obs_text: action_i", i may from 1~3...
action_index = [[end - start, end] for start, end in zip(self.action_list_length, sequence_length)]

# maybe no need to use it, directly use logits
logits = torch.log_softmax(outputs.logits, dim=-1)

logits = logits[:, :-1, :]
input_ids = input_ids[:, 1:]
gen_logits = torch.gather(logits, 2, input_ids[:, :, None]).squeeze(-1)

slices = [gen_logits[i, start-1:end-1] for i, (start, end) in enumerate(action_index)]

# extract the action logits from the sequence logits, just extract the index of [action_start_index, action_end_index] from the sequence logits
action_logits = torch.stack([torch.sum(s) for s in slices]) # sum the action logits as the final action_logits
```
I think this is an operation of extract the action logits from the sequence logits, based on the `action_token's start and end index` in sequence, and then sum the `action_token's logits` as the `action_logits`.

I don't understand why discarding other token's logits (like `obs_text's logits`). Will this not lead to the loss of information? Any possible to merge the other token's logits into the calculation of `action_logits`?

Thanks~

**Comment by WeihaoTan** [comment]
We only use the joint probabilities of action tokens to form the policy. The probabilities of these action tokens are conditioned on the obs_text. I do not think there is any loss of information.  And all the actions share the same obs prompt. Even if you add the probabilities of obs text, they will also be normalized during the softmax process. So I do not see any benefits of taking the probabilities of obs text into consideration. You could have a look at our paper for more details.

**Comment by Felixvillas** [comment]
> We only use the joint probabilities of action tokens to form the policy. The probabilities of these action tokens are conditioned on the obs_text. I do not think there is any loss of information. And all the actions share the same obs prompt. Even if you add the probabilities of obs text, they will also be normalized during the softmax process. So I do not see any benefits of taking the probabilities of obs text into consideration. You could have a look at our paper for more details.

Thanks for reply~ Nice works! I'll close this issue
